src/libtls/tls_server.c

```diff
@@ -167,71 +167,120 @@ struct private_tls_server_t {
 	bool curves_received;
 };
 
+/**
+ * Create an array of an intersection of server and peer supported key types
+ */
+static array_t *create_common_key_types(chunk_t hashsig,
+										tls_version_t version_min,
+										tls_version_t version_max)
+{
+	hashtable_t *ht;
+	array_t *key_types;
+	enumerator_t *enumerator;
+	int *k;
+	key_type_t *v, lookup;
+	uint16_t sig_scheme;
+
+	ht = tls_get_supported_key_types(version_min, version_max);
+	if (!ht)
+	{
+		return NULL;
+	}
+	key_types = array_create(sizeof(key_type_t), 1);
+	enumerator = ht->create_enumerator(ht);
+	while (enumerator->enumerate(enumerator, &k, &v))
```

Comment:
Without hashtable, you could simplify this to the following (also, even before the change to enumerator_t, the function always returned a valid pointer):

```suggestion
	array_t *key_types;
	enumerator_t *enumerator;
	key_type_t v, lookup;
	uint16_t sig_scheme;

	key_types = array_create(sizeof(key_type_t), 8);
	enumerator = tls_get_supported_key_types(version_min, version_max);
	while (enumerator->enumerate(enumerator, &v))